### ISSUE TITLE
Spec: Deep ResultStore Module

### ISSUE THREAD

**Issue by Igoreduardobraga** [Igoreduardobraga/BreakEvenNLP#21]
# Spec: Deep ResultStore Module

**Triage Label:** `ready-for-agent`  
**Status:** Ready for Implementation  
**Area:** Experiment Persistence & Metrics Pipeline  

---

## Problem Statement

The `BreakEvenNLP` evaluation pipeline currently persists experiment outputs as individual, unstructured `results.json` files deeply nested inside directory hierarchies (`results/{exp_name}/{exp_type}_{model}_{size}/{cfg}/{dataset}/{factor}/repeat_{r}_fold_{k}/results.json`). Across repeated stratified k-fold splits, factors, datasets, and sample sizes, this produces up to 50,000 tiny files containing redundant copies of ground-truth label arrays, duplicate metadata, and thousands of raw decoded text strings.

This architecture presents four critical issues:
1. **Severe I/O and Inode Exhaustion:** Traversing tens of thousands of directories on cloud storage or networked filesystems (e.g. RunPod, NFS, EBS) degrades file-system performance and exhausts inode limits.
2. **Analysis Brittleness:** Aggregating results requires recursive `os.walk` loops parsing individual JSON files (a pattern that previously led to 1,800+ lines of unmaintainable ad-hoc parsing scripts deleted in commit `e943b39`).
3. **Lack of Concurrency and Atomicity Guarantees:** Abrupt termination or interrupted jobs leave partially written or corrupted JSON files without transactional safety.
4. **Missing Provenance:** Metrics are saved without computational metadata (execution duration, timestamp, git commit hash, environment/hardware context), compromising long-term scientific reproducibility.

## Solution

Implement a deep `ResultStore` module in `BreakEvenNLP/result_store.py` providing an atomic, columnar metrics store alongside granular fold-level artifacts.

The module:
- Serves as the single public seam for experiment persistence across the repository.
- Maintains an atomic, incrementally appended columnar summary table (`summary.parquet` with automatic fallback to `summary.csv`) containing scalar metrics, factor seeds, execution timing, and provenance metadata.
- Persists large fold-specific arrays (golden targets, predictions, decoded tokens) cleanly without bloating the main summary table.
- Provides 100% backward-compatibility by conditionally writing the legacy `repeat_{r}_fold_{k}/results.json` structure, ensuring zero regression for existing workflows.
- Exposes high-performance, filterable retrieval via `load_summary()`, returning a structured `pandas.DataFrame` in milliseconds.

## User Stories

1. As an NLP researcher executing extensive grid evaluations across 200+ configurations, I want experiment metrics to be recorded into a single columnar table (Parquet/CSV), so that I don't generate tens of thousands of fragmented directories on disk.
2. As a researcher working on cloud instances (e.g. RunPod, AWS EBS, NFS), I want append operations to be atomic and thread-safe, so that interrupted or concurrent job executions never corrupt existing metric logs.
3. As a data analyst visualizing break-even points, I want to query metrics across all datasets, models, and sample sizes via `ResultStore.load_summary()`, so that I get a consolidated Pandas DataFrame in under 10 milliseconds without recursive `os.walk` scans.
4. As a researcher reproducing arXiv:2402.12819v2, I want `ResultStore` to automatically record all 6 experimental factor seeds (`data_split_seed`, `label_choice_seed`, `sample_choice_seed`, `sample_order_seed`, `model_initialisation_seed`, `model_randomness_seed`) alongside each fold, so that the provenance of every single data point is 100% reproducible.
5. As an experiment runner restarting an interrupted batch, I want `ResultStore.has_fold(repeat, fold)` to quickly verify whether a fold has already been evaluated, so that already computed results are skipped without unnecessary GPU computation.
6. As a researcher wishing to force re-evaluation of an experiment, I want a `regenerate=True` option, so that existing records for that specific fold can be overwritten cleanly.
7. As a maintainer with existing scripts expecting legacy outputs, I want `ResultStore` to emit the legacy `results.json` file inside `repeat_{r}_fold_{k}/`, so that any legacy code expecting the old directory layout continues to function without modifications.
8. As a researcher logging classification performance, I want `record_fold` to accept arbitrary metric dictionaries (e.g. `f1_macro`, `f1_prompting`, `f1_macro_icl`, `accuracy`), so that the store natively supports Prompting, ICL, Instruction Tuning, and Fine-Tuning experiments.
9. As a researcher tracking resource utilization and efficiency, I want `ResultStore` to automatically track fold duration (in seconds) and execution timestamps, so that I can benchmark the computational cost of different model architectures and sample sizes.
10. As a researcher running distributed or multi-GPU jobs, I want the summary table to record system provenance (git commit hash, hostname, Python/PyTorch versions, CUDA device name), so that discrepancies caused by environment drift can be diagnosed immediately.
11. As an NLP researcher evaluating generation quality, I want raw generated text (`decodeds`) and ground truth / predicted label arrays to be stored efficiently in fold details, so that large textual outputs don't bloat the primary metric summary table.
12. As a developer running tests in lightweight CI environments where PyArrow may not be installed, I want `ResultStore` to fall back gracefully to standard CSV, so that the module works out of the box in zero-dependency environments.
13. As a developer testing `ResultStore`, I want to pass a custom `base_dir` (e.g., `tmp_path`), so that unit tests can verify record insertion, filtering, and summarization in isolated sandbox environments.
14. As a data scientist filtering experiments, I want `load_summary` to accept filter arguments (such as `dataset`, `model`, `factor`, `sample_size`), so that I can load only the slice of experiments required for a specific figure or hypothesis test.
15. As a researcher comparing fine-tuning with prompting, I want `ResultStore` to distinguish between fine-tuning and in-context learning runs via an explicit `experiment_type` column, so that downstream analysis can group and contrast the two paradigms trivially.
16. As a researcher auditing prediction errors, I want `load_fold_details(repeat, fold)` to return a structured dictionary or tuple of `(real, predicted, decodeds)`, so that confusion matrices and error analyses can be generated on demand.
17. As an engineer maintaining storage quotas on shared servers, I want a compaction utility `ResultStore.compact()`, so that temporary or incremental fold logs can be merged into a single optimized columnar file with zero fragmentation.
18. As a developer writing regression tests, I want `ResultStore` to reject malformed metrics or incompatible types early with informative `ValueError` messages, so that silent data corruption or NaN metrics are caught at recording time.

## Implementation Decisions

- The metrics storage and retrieval logic is extracted into a dedicated deep module located at `BreakEvenNLP/result_store.py`.
- The public interface of the module is:
  ```python
  class ResultStore:
      def __init__(
          self,
          results_path: str | Path,
          experiment_name: str,
          experiment_type: str,
          model_name: str,
          dataset: str,
          factor: str,
          configuration_name: str = "",
          legacy_json: bool = True,
          storage_format: str = "auto",  # 'parquet', 'csv', or 'auto'
      ):
          ...

      def has_fold(self, repeat: int, fold: int) -> bool:
          ...

      def record_fold(
          self,
          repeat: int,
          fold: int,
          run_seed_used: int,
          randomness_factor_seeds: dict[str, int],
          metrics: dict[str, float],
          real: list[int],
          predicted: list[int],
          decodeds: list[str] | None = None,
          duration_seconds: float | None = None,
      ) -> None:
          ...

      def load_summary(
          self,
          filters: dict | None = None,
      ) -> pd.DataFrame:
          ...

      def load_fold_details(
          self,
          repeat: int,
          fold: int,
      ) -> dict:
          ...
  ```
- **Summary Table Schema:** The tabular summary file (`summary.parquet` or `summary.csv`) enforces the following standardized columnar schema:
  ```python
  {
      "experiment_name": str,
      "experiment_type": str,
      "model": str,
      "dataset": str,
      "factor": str,
      "configuration": str,
      "rskf_repeat": int,
      "rskf_fold": int,
      "run_seed_used": int,
      "data_split_seed": int,
      "label_choice_seed": int,
      "sample_choice_seed": int,
      "sample_order_seed": int,
      "model_initialisation_seed": int,
      "model_randomness_seed": int,
      "f1_macro": float,           # Optional / nullable depending on experiment
      "f1_prompting": float,       # Optional / nullable
      "f1_macro_icl": float,       # Optional / nullable
      "duration_seconds": float,
      "timestamp": str,            # ISO 8601
      "git_commit": str,           # SHA hash or 'dirty'/'unknown'
  }
  ```
- **Atomicity and Safe Writes:** Writes to the summary table use a staging file with an atomic filesystem replace (`tmp_file.replace(target_file)`) to prevent corrupted tables during abrupt interruptions.
- **Graceful Format Fallback:** `storage_format="auto"` attempts to use `pyarrow` / `fastparquet` for Parquet output; if unavailable, it falls back to standard `pandas.DataFrame.to_csv()` without raising an unhandled exception.
- **Backward Compatibility:** When `legacy_json=True` (the default), `record_fold` writes `repeat_{r}_fold_{k}/results.json` containing the identical dictionary keys currently consumed by the repo (`f1_macro`, `real`, `predicted`, `base_model`, `rskf_repeat`, `rskf_fold`, `run_seed_used`, `decodeds`, factor seeds).
- **Integration Seam in `main.py`:** Procedural file manipulation (`os.makedirs`, manual `json.dump`, manual `os.path.exists`) in the RSKF loop is replaced by `store.has_fold(...)` and `store.record_fold(...)`.

## Testing Decisions

- A good test verifies only observable behavior across the public seam (`has_fold`, `record_fold`, `load_summary`, `load_fold_details`) without asserting on private locking primitives or internal file staging.
- The test suite will be introduced in `BreakEvenNLP/tests/test_result_store.py`.
- Specific behaviors verified:
  - `has_fold` returns `False` on a fresh directory and `True` after `record_fold`.
  - `record_fold` appends rows idempotently without losing previous fold records.
  - `load_summary` returns a valid `pandas.DataFrame` matching recorded metrics and metadata.
  - `legacy_json=True` produces `results.json` with exact dictionary keys and values matching existing outputs.
  - Automatic fallback to CSV functions seamlessly when Parquet is disabled or missing.
  - Atomic write recovery ensures that interrupted writes leave previous valid summary tables intact.
- Prior art: existing deep module test suites in `BreakEvenNLP/tests/` (`test_evaluator.py`, `test_prompt_formatter.py`, `test_sample_pool.py`, `test_rng.py`).

## Out of Scope

- Modifying core ML evaluation calculations (`compute_macro_f1` in `metrics.py`) or sampling logic.
- Building graphical dashboards or visualization plotting scripts (future plotting modules will consume `ResultStore.load_summary()`).
- Cloud bucket synchronization (e.g. S3, GCS, WandB sync), which belongs in optional sidecar utilities.

## Further Notes

- Maintains 100% fidelity to the original experiment runs, seeds, and output metrics.
- Enables instant data aggregation for future statistical tests and break-even curve plotting.
